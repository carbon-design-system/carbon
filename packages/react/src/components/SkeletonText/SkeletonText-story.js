/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable no-console */

import React from 'react';

import { withKnobs, boolean, number, select } from '@storybook/addon-knobs';
import SkeletonText from '../SkeletonText';
import mdx from './SkeletonText.mdx';

const widths = {
  '100%': '100%',
  '250px': '250px',
};

const props = () => ({
  heading: boolean('Skeleton text at a larger size (heading)'),
  paragraph: boolean('Use multiple lines of text (paragraph)'),
  lineCount: number('The number of lines in a paragraph (lineCount)', 3),
  width: select(
    'Width (in px or %) of single line of text or max-width of paragraph lines (width)',
    widths,
    '100%'
  ),
});

export default {
  title: 'Components/Skeleton',
  decorators: [withKnobs],

  parameters: {
    component: SkeletonText,
    docs: {
      page: mdx,
    },
  },
};

export const _SkeletonText = () => (
  <div style={{ width: '300px' }}>
    <SkeletonText {...props()} />
  </div>
);
