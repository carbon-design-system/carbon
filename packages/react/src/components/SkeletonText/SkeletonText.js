/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import { usePrefix } from '../../internal/usePrefix';

const randoms = [0.973051493507435, 0.15334737213558558, 0.5671034553053769];

function getRandomInt(min, max, n) {
  return Math.floor(randoms[n % 3] * (max - min + 1)) + min;
}

const SkeletonText = ({
  paragraph,
  lineCount,
  width,
  heading,
  className,
  ...other
}) => {
  const prefix = usePrefix();
  const skeletonTextClasses = classNames({
    [`${prefix}--skeleton__text`]: true,
    [`${prefix}--skeleton__heading`]: heading,
    [className]: className,
  });

  const widthNum = parseInt(width, 10);

  const widthPx = width.includes('px');

  const widthPercent = width.includes('%');

  if (widthPercent && paragraph) {
    const lines = [];
    for (var i = 0; i < lineCount; i++) {
      const randomWidth = getRandomInt(0, 75, i) + 'px';
      lines.push(
        <p
          className={skeletonTextClasses}
          style={{ width: `calc(${width} - ${randomWidth})` }}
          key={i}
          {...other}
        />
      );
    }
    return <div>{lines}</div>;
  }

  if (widthPx && paragraph) {
    const lines = [];
    for (var j = 0; j < lineCount; j++) {
      const randomWidth = getRandomInt(widthNum - 75, widthNum, j) + 'px';
      lines.push(
        <p
          className={skeletonTextClasses}
          style={{ width: randomWidth }}
          key={j}
          {...other}
        />
      );
    }
    return <div>{lines}</div>;
  }
  return (
    <p className={skeletonTextClasses} style={{ width: width }} {...other} />
  );
};

SkeletonText.propTypes = {
  /**
   * Specify an optional className to be applied to the container node
   */
  className: PropTypes.string,
  /**
   * generates skeleton text at a larger size
   */
  heading: PropTypes.bool,
  /**
   * the number of lines in a paragraph
   */
  lineCount: PropTypes.number,
  /**
   * will generate multiple lines of text
   */
  paragraph: PropTypes.bool,
  /**
   * width (in px or %) of single line of text or max-width of paragraph lines
   */
  width: PropTypes.string,
};

SkeletonText.defaultProps = {
  paragraph: false,
  width: '100%',
  heading: false,
  lineCount: 3,
};

export default SkeletonText;
