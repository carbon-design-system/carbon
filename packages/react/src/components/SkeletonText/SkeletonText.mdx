import { Props } from '@storybook/addon-docs/blocks';

# SkeletonText

[Source code](https://github.com/carbon-design-system/carbon/tree/main/packages/react/src/components/SkeletonText)
&nbsp;|&nbsp;
[Usage guidelines](https://www.carbondesignsystem.com/patterns/loading-pattern/#skeleton-states)

## Table of Contents

## Overview

## Component API

<Props />

## Feedback

Help us improve this component by providing feedback, asking questions on Slack,
or updating this file on
[GitHub](https://github.com/carbon-design-system/carbon/edit/main/packages/react/src/components/SkeletonText/SkeletonText.mdx).
