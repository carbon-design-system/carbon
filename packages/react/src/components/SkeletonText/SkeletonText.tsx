/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import PropTypes from 'prop-types';
import React, { MutableRefObject, ReactNode, useRef } from 'react';
import classNames from 'classnames';
import { usePrefix } from '../../internal/usePrefix';
import useIsomorphicEffect from '../../internal/useIsomorphicEffect';

const randoms = [0.973051493507435, 0.15334737213558558, 0.5671034553053769];

function getRandomInt(min: number, max: number, n: number) {
  return Math.floor(randoms[n % 3] * (max - min + 1)) + min;
}

interface SkeletonTextProps {
  /**
   * Specify an optional className to be applied to the container node.
   */
  className?: string;

  /**
   * Generates skeleton text at a larger size.
   */
  heading?: boolean;

  /**
   * The number of lines shown if paragraph is true.
   */
  lineCount?: number;

  /**
   * Set this to true to generate multiple lines of text.
   */
  paragraph?: boolean;

  /**
   * Width (in px or %) of single line of text or max-width of paragraph lines.
   */
  width?: string;
}

const SkeletonText = ({
  paragraph = false,
  lineCount = 3,
  width = '100%',
  heading = false,
  className = '',
  ...rest
}: SkeletonTextProps) => {
  const prefix = usePrefix();
  const skeletonTextClasses = classNames({
    [`${prefix}--skeleton__text`]: true,
    [`${prefix}--skeleton__heading`]: heading,
    [className]: className,
  });

  const widthNum = parseInt(width, 10);
  const widthPx = width.includes('px');
  const widthPercent = width.includes('%');

  let lineCountNumber = 1;

  if (paragraph) {
    lineCountNumber = lineCount;
  }

  const refs: MutableRefObject<(HTMLParagraphElement | null)[]> = useRef([]);

  useIsomorphicEffect(() => {
    refs.current.map((item, j) => {
      const randomPercentWidth = getRandomInt(0, 75, j) + 'px';
      const randomPxWidth =
        getRandomInt(Math.max(widthNum - 75, 0), widthNum, j) + 'px';

      if (item) {
        if (widthPercent && paragraph) {
          item.style.width = `calc(${width} - ${randomPercentWidth})`;
        } else if (widthPx && paragraph) {
          item.style.width = randomPxWidth;
        } else {
          item.style.width = width;
        }
      }
    });
  }, [
    lineCountNumber,
    paragraph,
    refs,
    width,
    widthNum,
    widthPercent,
    widthPx,
  ]);

  const lines: ReactNode[] = [];
  for (let i = 0; i < lineCountNumber; i++) {
    lines.push(
      <p
        className={skeletonTextClasses}
        key={i}
        ref={(el) => (refs.current = [...refs.current, el])}
        {...rest}
      />
    );
  }

  if (lineCountNumber !== 1) {
    return <div>{lines}</div>;
  }

  // eslint-disable-next-line react/jsx-no-useless-fragment
  return <>{lines}</>;
};

SkeletonText.propTypes = {
  /**
   * Specify an optional className to be applied to the container node
   */
  className: PropTypes.string,
  /**
   * generates skeleton text at a larger size
   */
  heading: PropTypes.bool,
  /**
   * the number of lines shown if paragraph is true
   */
  lineCount: PropTypes.number,
  /**
   * will generate multiple lines of text
   */
  paragraph: PropTypes.bool,
  /**
   * width (in px or %) of single line of text or max-width of paragraph lines
   */
  width: PropTypes.string,
};

export default SkeletonText;
