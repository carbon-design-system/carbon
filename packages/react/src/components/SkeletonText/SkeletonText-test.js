/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import SkeletonText from '../SkeletonText';
import { shallow } from 'enzyme';

const prefix = 'cds';

describe('SkeletonText', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<SkeletonText />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass(`${prefix}--skeleton__text`)).toEqual(true);
    });
  });
});

describe('SkeletonText Heading', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<SkeletonText heading />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass(`${prefix}--skeleton__heading`)).toEqual(true);
    });
  });
});
