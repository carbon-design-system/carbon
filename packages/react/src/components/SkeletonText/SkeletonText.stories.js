/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable no-console */

import React from 'react';

import SkeletonText from '.';
import mdx from './SkeletonText.mdx';

export default {
  title: 'Components/Skeleton/SkeletonText',
  component: SkeletonText,
  parameters: {
    docs: {
      page: mdx,
    },
  },
};

export const Default = () => <SkeletonText />;

export const Playground = (args) => <SkeletonText {...args} />;

Playground.argTypes = {
  className: {
    control: false,
  },
  heading: {
    control: {
      type: 'boolean',
    },
    defaultValue: false,
  },
  paragraph: {
    control: {
      type: 'boolean',
    },
    defaultValue: false,
  },
  width: {
    control: {
      type: 'text',
    },
    defaultValue: '100%',
  },
  lineCount: {
    control: {
      type: 'number',
    },
    defaultValue: 3,
  },
};
