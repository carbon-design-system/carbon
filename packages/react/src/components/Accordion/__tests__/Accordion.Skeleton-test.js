/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { mount } from 'enzyme';
import React from 'react';
import AccordionSkeleton from '../Accordion.Skeleton';
import SkeletonText from '../../SkeletonText';

const prefix = 'cds';

describe('AccordionSkeleton', () => {
  it('should render', () => {
    const wrapper = mount(<AccordionSkeleton />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should support rendering without any open items', () => {
    const wrapper = mount(<AccordionSkeleton open={false} />);
    expect(wrapper.contains(<SkeletonText width="90%" />)).toEqual(false);
  });

  it('should support rendering a custom number of items', () => {
    const count = 8;
    const wrapper = mount(<AccordionSkeleton count={count} />);
    expect(wrapper.find(`.${prefix}--accordion__item`)).toHaveLength(count);
  });
});
