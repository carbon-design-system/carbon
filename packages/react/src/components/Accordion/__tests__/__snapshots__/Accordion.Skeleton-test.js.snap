// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionSkeleton should render 1`] = `
<AccordionSkeleton
  align="end"
  count={4}
  open={true}
>
  <ul
    className="bx--accordion bx--skeleton bx--accordion--end"
  >
    <li
      className="bx--accordion__item bx--accordion__item--active"
    >
      <span
        className="bx--accordion__heading"
      >
        <ForwardRef(ChevronRight16)
          className="bx--accordion__arrow"
        >
          <Icon
            className="bx--accordion__arrow"
            fill="currentColor"
            height={16}
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 16 16"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <svg
              aria-hidden={true}
              className="bx--accordion__arrow"
              fill="currentColor"
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
              />
            </svg>
          </Icon>
        </ForwardRef(ChevronRight16)>
        <SkeletonText
          className="bx--accordion__title"
          heading={false}
          lineCount={3}
          paragraph={false}
          width="100%"
        >
          <p
            className="bx--skeleton__text bx--accordion__title"
            style={
              Object {
                "width": "100%",
              }
            }
          />
        </SkeletonText>
      </span>
      <div
        className="bx--accordion__content"
      >
        <SkeletonText
          heading={false}
          lineCount={3}
          paragraph={false}
          width="90%"
        >
          <p
            className="bx--skeleton__text"
            style={
              Object {
                "width": "90%",
              }
            }
          />
        </SkeletonText>
        <SkeletonText
          heading={false}
          lineCount={3}
          paragraph={false}
          width="80%"
        >
          <p
            className="bx--skeleton__text"
            style={
              Object {
                "width": "80%",
              }
            }
          />
        </SkeletonText>
        <SkeletonText
          heading={false}
          lineCount={3}
          paragraph={false}
          width="95%"
        >
          <p
            className="bx--skeleton__text"
            style={
              Object {
                "width": "95%",
              }
            }
          />
        </SkeletonText>
      </div>
    </li>
    <AccordionSkeletonItem
      key="0"
    >
      <li
        className="bx--accordion__item"
      >
        <span
          className="bx--accordion__heading"
        >
          <ForwardRef(ChevronRight16)
            className="bx--accordion__arrow"
          >
            <Icon
              className="bx--accordion__arrow"
              fill="currentColor"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <svg
                aria-hidden={true}
                className="bx--accordion__arrow"
                fill="currentColor"
                focusable="false"
                height={16}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
                />
              </svg>
            </Icon>
          </ForwardRef(ChevronRight16)>
          <SkeletonText
            className="bx--accordion__title"
            heading={false}
            lineCount={3}
            paragraph={false}
            width="100%"
          >
            <p
              className="bx--skeleton__text bx--accordion__title"
              style={
                Object {
                  "width": "100%",
                }
              }
            />
          </SkeletonText>
        </span>
      </li>
    </AccordionSkeletonItem>
    <AccordionSkeletonItem
      key="1"
    >
      <li
        className="bx--accordion__item"
      >
        <span
          className="bx--accordion__heading"
        >
          <ForwardRef(ChevronRight16)
            className="bx--accordion__arrow"
          >
            <Icon
              className="bx--accordion__arrow"
              fill="currentColor"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <svg
                aria-hidden={true}
                className="bx--accordion__arrow"
                fill="currentColor"
                focusable="false"
                height={16}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
                />
              </svg>
            </Icon>
          </ForwardRef(ChevronRight16)>
          <SkeletonText
            className="bx--accordion__title"
            heading={false}
            lineCount={3}
            paragraph={false}
            width="100%"
          >
            <p
              className="bx--skeleton__text bx--accordion__title"
              style={
                Object {
                  "width": "100%",
                }
              }
            />
          </SkeletonText>
        </span>
      </li>
    </AccordionSkeletonItem>
    <AccordionSkeletonItem
      key="2"
    >
      <li
        className="bx--accordion__item"
      >
        <span
          className="bx--accordion__heading"
        >
          <ForwardRef(ChevronRight16)
            className="bx--accordion__arrow"
          >
            <Icon
              className="bx--accordion__arrow"
              fill="currentColor"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <svg
                aria-hidden={true}
                className="bx--accordion__arrow"
                fill="currentColor"
                focusable="false"
                height={16}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 16 16"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
                />
              </svg>
            </Icon>
          </ForwardRef(ChevronRight16)>
          <SkeletonText
            className="bx--accordion__title"
            heading={false}
            lineCount={3}
            paragraph={false}
            width="100%"
          >
            <p
              className="bx--skeleton__text bx--accordion__title"
              style={
                Object {
                  "width": "100%",
                }
              }
            />
          </SkeletonText>
        </span>
      </li>
    </AccordionSkeletonItem>
  </ul>
</AccordionSkeleton>
`;
